src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -71,21 +71,34 @@ public override Task DocumentOpenAsync(Document document, CancellationToken canc
             }
         }
 
+        public override Task DocumentCloseAsync(Document document, CancellationToken cancellationToken)
+        {
+            using (Logger.LogBlock(FunctionId.Diagnostics_DocumentClose, GetResetLogMessage, document, cancellationToken))
+            {
+                return DocumentResetOrCloseAsync(document, raiseEvent: true, cancellationToken: cancellationToken);
+            }
+        }
+
         public override Task DocumentResetAsync(Document document, CancellationToken cancellationToken)
         {
             using (Logger.LogBlock(FunctionId.Diagnostics_DocumentReset, GetResetLogMessage, document, cancellationToken))
             {
-                // we don't need the info for closed file
-                _memberRangeMap.Remove(document.Id);
-
-                // we remove whatever information we used to have on document open/close and re-calculate diagnostics
-                // we had to do this since some diagnostic analyzer changes its behavior based on whether the document is opened or not.
-                // so we can't use cached information.
-                ClearDocumentStates(document, _stateManger.GetStateSets(document.Project), raiseEvent: false, cancellationToken: cancellationToken);
-                return SpecializedTasks.EmptyTask;
+                return DocumentResetOrCloseAsync(document, raiseEvent: false, cancellationToken: cancellationToken);
```

Comment:
Can you add some explanatory comment about why raiseEvent is true for DocumentClose but false for DocumentReset?
I guess you are doing so because we want to clear the diagnostics from error list for DocumentClose, but refresh with re-computed diagnostics from DocumentReset?